### PR TITLE
fix(ci): gate PR-review selection on the required checks, not every check (#1086)

### LINKED ISSUE #1086: The PR-review selector skips on ANY failing check, so an advisory red blocks the approval the merge requires
## What happens

`.github/scripts/pr_review_select.py` refuses to select a pull request for automated review when **any** check-run on its head has failed:

```python
green, why = checks_green(repo, sha)
if not green:
    skipped.append((number, why))
    continue
```

The script contains **no** notion of required versus optional checks — `grep -c 'REQUIRED_CHECKS\|required'` over it returns **0**. Branch protection, meanwhile, requires exactly `build`, `build-macos`, `build-windows`, `security` and `e2e-deterministic` (`scripts/apply-branch-protection.sh:128`).

So a job that branch protection deliberately treats as **advisory** still blocks reviewer selection outright.

## Why that matters more than it sounds

The advisory job cannot block the merge button. But the automated reviewer's approval **is** what satisfies the required-review rule, so an advisory failure blocks the only thing that can unblock the merge. **"Advisory" is therefore not advisory** — it is load-bearing through a second path nobody looks at.

The two statuses are also easy to confuse from the outside: the `review` and `vote` jobs come back `skipped`, not `failure`, and the reason lives only in the `prepare` job's log:

```
skip #1076: check 'devbox' concluded 'failure'
selected 0, live=True
##[warning]#1076 was requested but is not eligible; see skip reasons above
```

From the pull request page that is indistinguishable from the reviewer having no quota — which is the failure mode the review workflow's own bounding instructions are written around (*"an exhausted quota produces no check-run at all, indistinguishable from the app being uninstalled"*). Here the cause is fully known and logged, but only if you go and read the `prepare` job.

## Observed

PRD #742's PR ([#1076](https://github.com/vfarcic/dot-agent-deck/pull/1076)) had **all five required checks green** and one advisory job red — `devbox`, which died 15–18 seconds in on `curl: (22) ... 504` from `https://artifacts.nixos.org/experimental-installer` inside the third-party `jetify-com/devbox-install-action`. The diff touched zero files under `devbox.json`, `devbox.lock`, `flake.nix`, `flake.lock` or `.github/workflows/`, and the sibling `nix` job passed on the same run.

An external CDN outage on a job that does not gate the merge therefore withheld the approval that does.

## Suggested fix

Have the selector's check gate consider **only the checks branch protection requires**, rather than every check-run on the head. Both halves of that pair already exist in the repository — `REQUIRED_CHECKS` in `scripts/apply-branch-protection.sh` is the list — so the fix is to make the selector read the same source of truth rather than inventing a stricter rule of its own.

Worth deciding at the same time:

- **Should an advisory red be surfaced to the reviewer as context rather than as a gate?** A reviewer that knows `devbox` is failing might reasonably mention it; one that refuses to run says nothing to anyone.
- **Should a skip-for-ineligibility be visible on the pull request** rather than only in the `prepare` log? Today a requester sees the same "nothing happened" as a quota exhaustion, and has to know which job to open to tell them apart.

## Ordering

**Independent of PRD #742's PR — before or after, either way.** It touches CI tooling only, and nothing in that PR's diff.

## Evidence

Found while settling PR #1076: the reviewer was dispatched, completed in about a minute without voting, and the `prepare` log named the reason.

### PR DESCRIPTION
**Pull request by vfarcic**
Fixes #1086.

## What was wrong

`.github/scripts/pr_review_select.py` refused a pull request whenever **any** check-run on its head had failed, required or not. Branch protection requires five contexts and leaves everything else advisory, so an advisory red cannot hold the merge button — but the reviewer's approval is what satisfies the required-review rule, so it held the only thing that could release the merge. PR #1076 lost its approval to a `devbox` job that died fifteen seconds in on a `curl: (22) … 504` from a third-party CDN inside someone else's action, on a diff touching no devbox or workflow file, with all five required contexts green.

## What changed

The gate is now the contexts branch protection **requires**, and that list is **derived at run time** from the live ruleset rather than hardcoded a third time.

- `required_contexts(repo)` reads `GET /repos/{repo}/rules/branches/{default_branch}` and unions the `required_status_checks` contexts, memoised per process — two API calls per run, not two per pull request.
- `classify_check_runs(runs, required)` is a new **pure** function: it decides green/not-green from a list of check runs, and returns the advisory failures it ignored. All the new tests drive this directly, with no network.
- `check_status(repo, sha)` fetches and delegates. `checks_green(repo, sha)` keeps its **two-value** contract on top of it.

### On the source of truth

The issue cites `scripts/apply-branch-protection.sh:128`. I confirmed that is still where the applier's list lives, and that it is not where enforcement lives: the legacy branch-protection API returns `required_status_checks: []` with `strict: null`, while the `main-protected` ruleset (id `20587589`) reports `build build-macos build-windows security e2e-deterministic`. Measured 2026-09-15, both agreeing with each other and with the script.

So rather than adding a third copy, the reviewer reads the ruleset. That is not only tidiness — it is what makes the narrower gate **defensible rather than merely looser**. The blanket "any red blocks" rule existed as a stated fail-closed hedge against the ruleset requiring a context `pr_review_common.py` did not list; a list re-read from the ruleset on every run is not a list anyone has to remember to update, so the hedge is answered rather than accepted.

The hardcoded five survive only as `FALLBACK_REQUIRED_CONTEXTS`, used when the ruleset cannot be read. That path prints a `::warning::` and a note naming the fallback list, and `the_fallback_list_matches_the_branch_protection_script` pins the tuple against the applier script's `REQUIRED_CHECKS` default so a drift between the two goes red in `cargo test-fast`.

### Skip reasons stayed legible, and gained a line

The issue's second complaint was that the reason was invisible. Nothing was removed:

- a skip still names the required context that held the pull request (`skip #1093: required context 'build-windows' concluded 'failure'`);
- a selected pull request with an advisory red now says so (`note #1076: advisory (not required, not a gate): devbox=failure`);
- the derived list is printed once per run (`required contexts, from the 'main' ruleset: [...]`), so a reader debugging a skip can see what the gate was.

The two open questions the issue raised are deliberately **not** answered here. Plumbing advisory reds into the agent's prompt needs a new input on the generated `pr-review.lock.yml`; posting a skip notice onto the pull request needs `pull-requests: write` in the `prepare` job, which today has `contents: read`. Both are scope expansions on a workflow that is deny-listed for a reason; they belong in their own change if wanted.

## The one judgement call — please check it

The instructions said not to loosen what the **vote** job validates. I kept every validation the vote job performs: `pr_review_vote.py` is **untouched**, it still re-fetches the head's check runs itself rather than trusting the selector, and it still re-validates the PR number, the head SHA, deny paths and its own per-head idempotence.

What does move under it is the shared gate **policy**, because `checks_green` is the same function. That is deliberate, and I think the alternative is worse:

- the two jobs are in one module precisely so they cannot apply different policies (the module docstring says so, and `DENY_PRECEDENCE` exists because they once did);
- if selection loosened and voting stayed strict, the reviewer would produce a verdict and the vote job would then refuse to cast it — the approval would still be withheld on an advisory red, which is exactly the bug #1086 is about. The fix would do nothing.

`the_selector_and_the_vote_job_share_one_check_gate` pins both halves: `pr_review_vote.checks_green is pr_review_common.checks_green`, and `checks_green` still returns exactly two values.

## Tests

Twelve new tests in `xtask/linkage-check/src/pr_review_verdict.rs`, the established pattern (CLAUDE.md rule 5's documented exceptions: `verify_pr_stream.rs`, `junit_strip.rs`, `issue_labeler_memory.rs`). They drive the real Python module under `python3`, feed synthetic check lists, and skip with `SKIP:` where `python3` is absent. All 42 tests in that module run in ~0.17s.

They are mutation-checked rather than assumed:

- restoring the old blanket "any failed check" loop turns **4** of them red (`an_advisory_failure_no_longer_blocks_selection`, `every_bad_conclusion_on_an_advisory_job_is_still_advisory`, `the_gate_follows_the_derived_list_not_a_hardcoded_one`, `the_ignored_advisory_reds_are_named_in_the_reason`);
- drifting one name in `FALLBACK_REQUIRED_CONTEXTS` turns `the_fallback_list_matches_the_branch_protection_script` red.

Also exercised end to end against the live repository — `pr_review_select.py` in `review` mode, dry, no writes — which derived the five contexts from the ruleset, skipped #1093 on a genuinely failing `build-windows`, skipped #612 as a draft, and selected #1070.

## Gates run

- `cargo fmt --check` — clean
- `cargo clippy --workspace --all-targets --features e2e,e2e-live -- -D warnings` — clean
- `cargo test-fast` — 3106 passed, 0 failed
- `cargo nextest run -p xtask-linkage-check pr_review_verdict` — 42 passed
- `cargo xtask linkage-check` — ok (12 rules)

Per CLAUDE.md rule 5 the full e2e tier was not run locally: lane 1 runs here in CI as `e2e-deterministic`, and lane 2 does not apply — there is no real-agent path in this change.

**Rule 12 does not apply**: no daemon, TUI↔daemon protocol, orchestration or hook surface is touched, so there is no `PROTOCOL_VERSION` bump and no cross-version manual test. Changelog fragment is `changelog.d/1086.misc.md` (CI/CD tooling).

## Known limitation

`pr-review-batch.yml`'s `prepare` job checks out the **base** branch, so this code cannot be exercised by a workflow run until it is on `main`. The one thing that therefore stays unverified until then is whether the job's `GITHUB_TOKEN` (`permissions: contents: read`) can read the rules endpoint. If it cannot, the run falls back to the hardcoded five with a `::warning::` naming the failure — i.e. it degrades to today's required-context behaviour and says so, rather than failing or going quiet.

## Review round 1 — both findings fixed (c1cc9be3)

Greptile raised two, both valid, and one of them found a **live bug older than this PR**.

**P2 (`pr_review_verdict.rs`) — re-run ordering was asserted and backwards.** `checks_green` kept whichever entry for a name came last in the check-runs response, under a comment claiming "re-runs append". Measured on `4afd7498`: `GET /commits/{sha}/check-runs` returns **newest first**, with all six of that head's duplicated names carrying their later `started_at` and higher `id` at the lower index (11-16 newer, 17-22 older, no exception). The old loop therefore kept the **oldest** run every time, so a check re-run from red to green went on reading as red — #1086's own symptom by another route. `_newest_key` now orders by `(started_at, id)` and the fetch selects both, so nothing depends on how the endpoint sorts.

**P1 (`pr_review_common.py`) — an empty ruleset is now taken at its word.** Falling back to the hardcoded five when the ruleset requires nothing reintroduced this issue one level up. The fallback is now reserved for a read that *failed*. The cost of believing an empty list is that a head with a failing build becomes eligible, so it is announced with a `::warning::` naming the remedy rather than passed over — the same answer `ALLOW_NO_REQUIRED_CHECKS` gives in the applier script for the same silent-weakening risk.

Five more tests, each mutation-checked against the pre-fix behaviour, including the derivation itself with `gh` stubbed in-process (empty / populated / multi-ruleset union / failed read / memoised to one read per run). Re-run gates: `cargo fmt --check` clean, the four-flag clippy clean, `cargo test-fast` 3111 passed, `cargo xtask linkage-check` ok, and `pr_review_verdict` now 47 tests.

Both threads answered and resolved.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: .github/scripts/pr_review_common.py [hunk 1]
```diff
@@ -189,32 +198,153 @@ def gh_json_paginated(*args):
...
+                        " | .parameters.required_status_checks[]? | .context]",
+            )
+            # Several rulesets can apply to one branch; the union is what the
+            # merge button waits for. Order is fixed so the log line is stable.
+            names = sorted({c for c in rules if c})
+            if names:
+                contexts = tuple(names)
+                # Said out loud once per run: the gate is now data rather than a
+                # constant, so a reader debugging a skip needs to see what it was.
+                print(f"required contexts, from the {target!r} ruleset: {list(contexts)}")
```
[the remark this answers is not in the document]
Comment (vfarcic) [reply .github/scripts/pr_review_common.py]: Agreed, and fixed in c1cc9be3 — you are right that the fallback was answering two different questions with one branch, and that the "authoritative empty" answer it gave was the one this PR exists to remove. A reviewer holding out for checks GitHub no longer requires is #1086 one level up.

The fallback is now reserved for a read that **failed**, which is the one case a reader genuinely cannot distinguish from "nothing is required". An empty derived tuple is returned as-is.

One thing I did not want to lose from the old branch, so it is kept as a warning rather than as a stricter gate. The cost of believing an empty list is that a head with a failing `build` becomes eligible for an approving vote, and an empty required list is easy to arrive at by accident here — `scripts/apply-branch-protection.sh` warns at length that `apply` sends a full `PUT`, so omitting the rule strips every required context "with no error, no warning and a success exit", and makes that path take an affirmative `ALLOW_NO_REQUIRED_CHECKS=true` for exactly that reason. So the weakening is now announced with the same remedy that script uses — loud rather than hidden:

```
::warning::branch protection requires no status checks on 'main', so the review gate
is checking none — a head with a failing build is now eligible. Restore the contexts
with scripts/apply-branch-protection.sh if this was not deliberate.
```

Covered by `the_required_list_comes_from_the_ruleset_and_falls_back_only_on_failure`, which stubs `gh` in-process and pins all four cases: empty → `()` plus the warning and no fallback note; populated → sorted union across several rulesets; failed read → `FALLBACK_REQUIRED_CONTEXTS` plus `::warning::could not read`. Mutation-checked — restoring `or None` on the empty case turns it red.

File: xtask/linkage-check/src/pr_review_verdict.rs [hunk 2]
```diff
@@ -547,3 +556,211 @@ fn an_unrejected_pull_request_is_not_a_stale_rejection() {
...
+
+/// Re-runs append rather than replace, so the LAST entry for a name is the
+/// current one. A required context re-run to green must read as green, and an
+/// advisory one re-run to green must drop out of the advisory list.
+#[test]
+fn a_re_run_check_is_read_at_its_newest_conclusion() {
+    assert_py_ok(
+        "runs = [crun('build', 'failure')] + green_five() + \\\n\
+        \x20   [crun('devbox', 'failure'), crun('devbox')]\n\
+         green, why, advisory = classify_check_runs(runs, REQ)\n\
```
[the remark this answers is not in the document]
Comment (vfarcic) [reply xtask/linkage-check/src/pr_review_verdict.rs]: Confirmed, and it was worse than unverified — the assertion was **backwards**, and it is a live bug rather than a latent one. Fixed in c1cc9be3. Thank you for pushing on this; I had carried the comment forward from the existing code without measuring it, which is precisely the rule 17 failure you cite.

Measured on `4afd7498`, a head where the gh-aw workflow produced two runs of the same six check names. `GET /commits/{sha}/check-runs` returns **newest first** — all six duplicated names carry their later `started_at` and higher `id` at the **lower** index, indices 11–16 holding the newer run of each pair and 17–22 the older, with no exception:

```
activation:        index 16 @ 15:11:13Z   index 22 @ 15:11:12Z
agent:             index 15 @ 15:11:14Z   index 21 @ 15:11:12Z
conclusion:        index 11 @ 15:11:14Z   index 17 @ 15:11:13Z
detection:         index 14 @ 15:11:14Z   index 20 @ 15:11:13Z
push_repo_memory:  index 13 @ 15:11:14Z   index 19 @ 15:11:13Z
safe_outputs:      index 12 @ 15:11:14Z   index 18 @ 15:11:13Z
```

So `by_name[run["name"]] = run` — last write wins — kept the **oldest** run for every duplicated name. A check re-run from red to green went on reading as red, which on a required context is this issue's own symptom arrived at by a different route.

The fix does not swap the positional rule for the opposite positional rule, which would just be the same unverified assumption mirrored. `_newest_key` orders by `(started_at, id)` and `check_status` now selects both fields, so the outcome does not depend on how the endpoint sorts or on it continuing to. `started_at` is ISO-8601 UTC and sorts lexicographically; `id` breaks a tie and rises with creation; a record carrying neither sorts oldest, so a partial or synthetic entry cannot displace a real one.

Three tests, all mutation-checked against the pre-fix loop:

- `a_re_run_check_is_read_at_its_newest_conclusion` — feeds the pair in **both** orders and asserts the same answer each way, in both the red→green and green→red directions, so it fails against a positional rule pointing either way.
- `a_re_run_advisory_check_drops_out_once_it_is_green` — a `devbox` re-run to green leaves the advisory list instead of being reported off its stale red.
- `the_newest_run_is_pinned_by_id_then_by_nothing_at_all` — the `id` tiebreak, and the partial-record case.

**Review by prageethw — APPROVED** [review]
Reviewed the full diff. The gate now reads required contexts from the live ruleset at run time (with a fallback pinned against apply-branch-protection.sh's default), fixes the re-run-ordering bug (API returns newest-first, old code kept the oldest), and reports advisory failures rather than treating them as a veto — well-motivated by the PR #1076 incident. 16 new unit tests cover every branch including the empty-ruleset edge case Greptile flagged, and that finding plus the other resolved thread are both addressed. dot-agent-deck-reviewer approved.
